src/build_helper/src/git.rs

```diff
@@ -27,145 +28,195 @@ pub fn output_result(cmd: &mut Command) -> Result<String, String> {
-/// upstream        https://github.com/rust-lang/rust (fetch)
-/// upstream        https://github.com/rust-lang/rust (push)
-/// ```
-pub fn get_rust_lang_rust_remote(
+/// Represents the result of checking whether a set of paths
+/// have been modified locally or not.
+#[derive(PartialEq, Debug, Clone)]
+pub enum PathFreshness {
+    /// Artifacts should be downloaded from this upstream commit,
+    /// there are no local modifications.
+    LastModifiedUpstream { upstream: String },
+    /// There are local modifications to a certain set of paths.
+    /// "Local" essentially means "not-upstream" here.
+    /// `upstream` is the latest upstream merge commit that made modifications to the
+    /// set of paths.
+    HasLocalModifications { upstream: String },
+    /// No upstream commit was found.
+    /// This should not happen in most reasonable circumstances, but one never knows.
+    MissingUpstream,
+}
+
+/// This function figures out if a set of paths was last modified upstream or
+/// if there are some local modifications made to them.
+/// It can be used to figure out if we should download artifacts from CI or rather
+/// build them locally.
+///
+/// The function assumes that at least a single upstream bors merge commit is in the
+/// local git history.
+///
+/// `target_paths` should be a non-empty slice of paths (git `pathspec`s) relative to `git_dir`
+/// whose modifications would invalidate the artifact.
+/// Each pathspec can also be a negative match, i.e. `:!foo`. This matches changes outside
+/// the `foo` directory.
+/// See https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefpathspecapathspec
+/// for how git `pathspec` works.
+///
+/// The function behaves differently in CI and outside CI.
+///
+/// - Outside CI, we want to find out if `target_paths` were modified in some local commit on
+/// top of the latest upstream commit that is available in local git history.
+/// If not, we try to find the most recent upstream commit (which we assume are commits
+/// made by bors) that modified `target_paths`.
+/// We don't want to simply take the latest master commit to avoid changing the output of
+/// this function frequently after rebasing on the latest master branch even if `target_paths`
+/// were not modified upstream in the meantime. In that case we would be redownloading CI
+/// artifacts unnecessarily.
+///
+/// - In CI, we use a shallow clone of depth 2, i.e., we fetch only a single parent commit
+/// (which will be the most recent bors merge commit) and do not have access
+/// to the full git history. Luckily, we only need to distinguish between two situations:
+/// 1) The current PR made modifications to `target_paths`.
+/// In that case, a build is typically necessary.
+/// 2) The current PR did not make modifications to `target_paths`.
+/// In that case we simply take the latest upstream commit, because on CI there is no need to avoid
+/// redownloading.
```

Comment:
Am I understanding right that `git` is able to distinguish with a shallow commit history between "the commits I can see" and "stuff beyond the horizon"? My mental model has always been that the first commit in a shallow repo looks as-if it had everything, but maybe that's wrong?

(Under that model, the two cases should be the same: the last commit to have modified stuff is either a local modification, and so we can't download, or it is that 'aggregate' commit).

Reply:
We don't know what exactly git does here, but the output of `git log -n 1 <filename>` for shallow clones seems pretty arbitrary in terms of which commit it assigns a change to; see https://github.com/rust-lang/rust/issues/101907#issuecomment-2697671282.

Reply:
I don't think there are problems per-se -- this is sort of referencing the other adjacent thread. I'd personally prefer if we could not have so much special casing for CI but I do think we need at least *something* specifically for the `auto` branch (shouldn't be needed on PR CI). I'll go ahead and resolve this.